src/net-questdb-client/Ingress/Senders/HttpSender.cs

```diff
@@ -0,0 +1,533 @@
+                            chain.ChainPolicy.CustomTrustStore.Add(
+                                X509Certificate2.CreateFromPemFile(Options.tls_roots, Options.tls_roots_password));
+                        }
+
+                        return chain!.Build(new X509Certificate2(certificate!));
+                    };
+            }
+
+            if (!string.IsNullOrEmpty(Options.tls_roots))
+            {
+                _handler.SslOptions.ClientCertificates ??= new X509Certificate2Collection();
+                _handler.SslOptions.ClientCertificates.Add(
+                    X509Certificate2.CreateFromPemFile(Options.tls_roots!, Options.tls_roots_password));
+            }
+        }
+
+        _handler.ConnectTimeout = Options.auth_timeout;
+        _handler.PreAuthenticate = true;
+
+        _client = new HttpClient(_handler);
+        var uri = new UriBuilder(Options.protocol.ToString(), Options.Host, Options.Port);
+        _client.BaseAddress = uri.Uri;
+        _client.Timeout = Timeout.InfiniteTimeSpan;
+
+        if (!string.IsNullOrEmpty(Options.username) && !string.IsNullOrEmpty(Options.password))
+        {
+            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
+                Convert.ToBase64String(Encoding.ASCII.GetBytes($"{Options.username}:{Options.password}")));
+        }
+        else if (!string.IsNullOrEmpty(Options.token))
+        {
+            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", Options.token);
+        }
+
+        return this;
+    }
+
+    /// <summary>
+    ///     Creates a new HTTP request with appropriate encoding and timeout.
+    /// </summary>
+    /// <returns></returns>
+    private (HttpRequestMessage, CancellationTokenSource?) GenerateRequest(CancellationToken ct = default)
+    {
+        var request = new HttpRequestMessage(HttpMethod.Post, "/write")
+            { Content = new BufferStreamContent(_buffer) };
+        request.Content.Headers.ContentType = new MediaTypeHeaderValue("text/plain") { CharSet = "utf-8" };
+        request.Content.Headers.ContentLength = _buffer.Length;
+        var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
+        cts.CancelAfter(CalculateRequestTimeout());
+        return (request, cts);
+    }
+
+    /// <summary>
+    ///     Calculate the request timeout.
+    /// </summary>
+    /// <remarks>
+    ///     Large requests may need more time to transfer the data.
+    ///     This calculation uses a base timeout (<see cref="QuestDBOptions.request_timeout" />), and adds on
+    ///     extra time corresponding to the expected transfer rate (<see cref="QuestDBOptions.request_min_throughput" />)
+    /// </remarks>
+    /// <returns></returns>
+    private TimeSpan CalculateRequestTimeout()
+    {
+        return Options.request_timeout
+               + TimeSpan.FromSeconds(_buffer.Length / (double)Options.request_min_throughput);
```

Comment:
Please don't change the request timeout from request to request, it should be constant regardless of the request size.